gateway/webhook/integration_test.go

```diff
@@ -0,0 +1,361 @@
+	gwURL := fmt.Sprintf("http://localhost:%d", httpPort)
+
+	require.Eventually(t, func() bool {
+		resp, err := http.Get(fmt.Sprintf("%s/health", gwURL))
+		if err != nil {
+			return false
+		}
+		defer resp.Body.Close()
+		return resp.StatusCode == http.StatusOK
+	}, time.Millisecond*500, time.Millisecond)
+
+	for _, tc := range testSetup.cases {
+		writeKey := tc.config.WriteKey
+		sourceID := tc.config.ID
+		workspaceID := tc.config.WorkspaceID
+
+		t.Run(tc.Description, func(t *testing.T) {
+			t.Logf("writeKey: %s", writeKey)
+			t.Logf("sourceID: %s", sourceID)
+			t.Logf("workspaceID: %s", workspaceID)
+
+			t.Log("Request Body:", string(tc.Input.Request.Body))
+
+			query, err := url.ParseQuery(tc.Input.Request.RawQuery)
+			require.NoError(t, err)
+			query.Set("writeKey", writeKey)
+
+			t.Log("Request URL:", fmt.Sprintf("%s/v1/webhook?%s", gwURL, query.Encode()))
+			req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("%s/v1/webhook?%s", gwURL, query.Encode()), bytes.NewBuffer(tc.Input.Request.Body))
+			require.NoError(t, err)
+
+			req.Header.Set("X-Forwarded-For", testSetup.context.RequestIP)
+			for k, v := range tc.Input.Request.Headers {
+				req.Header.Set(k, v)
+			}
+
+			resp, err := http.DefaultClient.Do(req)
+			require.NoError(t, err)
+			defer resp.Body.Close()
+
+			b, err := io.ReadAll(resp.Body)
+			require.NoError(t, err)
+
+			assert.Equal(t, tc.Output.Response.StatusCode, resp.StatusCode)
+			if resp.Header.Get("Content-Type") == "application/json" {
+				assert.JSONEq(t, string(tc.Output.Response.Body), string(b))
+			} else {
+				assert.Equal(t, string(tc.Output.Response.Body), fmt.Sprintf("%q", b))
+			}
+			r, err := gatewayDB.GetUnprocessed(ctx, jobsdb.GetQueryParams{
+				WorkspaceID: workspaceID,
+				ParameterFilters: []jobsdb.ParameterFilterT{{
+					Name:  "source_id",
+					Value: sourceID,
+				}},
+				JobsLimit: 10,
+			})
+			require.NoError(t, err)
+
+			assert.Len(t, r.Jobs, len(tc.Output.Queue), "enqueued items mismatch")
+			for i, p := range tc.Output.Queue {
+				var batch struct {
+					Batch []json.RawMessage `json:"batch"`
+				}
+				err := json.Unmarshal(r.Jobs[i].EventPayload, &batch)
+				require.NoError(t, err)
+				assert.Len(t, batch.Batch, 1)
+
+				if gjson.GetBytes(p, "messageId").String() == uuid.Nil.String() {
+					rawMsgID := gjson.GetBytes(batch.Batch[0], "messageId").String()
+					msgID, err := uuid.Parse(rawMsgID)
+					assert.NoErrorf(t, err, "messageId (%q) is not a valid UUID", rawMsgID)
+
+					p, err = sjson.SetBytes(p, "messageId", msgID.String())
+					require.NoError(t, err)
+				}
+
+				assert.JSONEq(t, string(p), string(batch.Batch[0]))
+			}
+
+			r, err = errDB.GetUnprocessed(ctx, jobsdb.GetQueryParams{
+				WorkspaceID: workspaceID,
+				// ParameterFilters: []jobsdb.ParameterFilterT{{
+				// 	Name:  "source_id",
+				// 	Value: sourceID,
+				// }},
+				JobsLimit: 1,
+			})
+			require.NoError(t, err)
+			assert.Len(t, r.Jobs, len(tc.Output.ErrQueue))
```

Comment:
Out of curiosity, how come sometimes you use `require` and sometimes `assert`? :thinking:

Reply:
I wanted to start introducing assert in my test to make the failed output a bit more useful, by having multiple assert logs instead of a single one. I kept the `require` for more severe errors. I understand the readability issue from inconsistency, need to reconsider this.